src/TreeSupport.cpp

```diff
@@ -0,0 +1,468 @@
+    const double diameter_angle_scale_factor = sin(storage.getSettingInAngleRadians("support_tree_branch_diameter_angle")) * layer_height / branch_radius; //Scale factor per layer to produce the desired angle.
+    const coord_t radius_sample_resolution = storage.getSettingInMicrons("support_tree_collision_resolution");
+    const bool support_rests_on_model = storage.getSettingAsSupportType("support_type") == ESupportType::EVERYWHERE;
+    for (size_t layer_nr = contact_nodes.size() - 1; layer_nr > 0; layer_nr--) //Skip layer 0, since we can't drop down the vertices there.
+    {
+        //Group together all nodes for each part.
+        std::vector<PolygonsPart> parts = model_avoidance[0][layer_nr].splitIntoParts();
+        std::vector<std::unordered_map<Point, Node>> nodes_per_part;
+        nodes_per_part.emplace_back(); //All nodes that aren't inside a part get grouped together in the 0th part.
+        for (size_t part_index = 0; part_index < parts.size(); part_index++)
+        {
+            nodes_per_part.emplace_back();
+        }
+        for (Node node : contact_nodes[layer_nr])
+        {
+            if (node.to_buildplate)
+            {
+                nodes_per_part[0][node.position] = node;
+            }
+            else if (support_rests_on_model)
+            {
+                //Find which part this node is located in and group the nodes in the same part together.
+                for (size_t part_index = 0; part_index < parts.size(); part_index++)
+                {
+                    if (parts[part_index].inside(node.position))
+                    {
+                        nodes_per_part[part_index + 1][node.position] = node;
+                        break;
+                    }
+                }
+            }
+        }
+        //Create a MST for every part.
+        std::vector<MinimumSpanningTree> spanning_trees;
+        for (const std::unordered_map<Point, Node> group : nodes_per_part)
+        {
+            std::unordered_set<Point> points_to_buildplate;
+            for (std::pair<Point, Node> entry : group)
+            {
+                points_to_buildplate.insert(entry.first); //Just the position of the node.
+            }
+            spanning_trees.emplace_back(points_to_buildplate);
+        }
+
+        for (size_t group_index = 0; group_index < nodes_per_part.size(); group_index++)
+        {
+            const MinimumSpanningTree& mst = spanning_trees[group_index];
+            //In the first pass, merge all leaf nodes.
+            for (std::pair<Point, Node> entry : nodes_per_part[group_index])
+            {
+                const Node node = entry.second;
+                std::vector<Point> neighbours = mst.adjacentNodes(node.position);
+                if (neighbours.size() == 1 && vSize2(neighbours[0] - node.position) < maximum_move_distance * maximum_move_distance) //This leaf is about to collapse. Merge it with the neighbour.
```

Comment:
We have to find more ways to lose points. Maybe if a node has 2 neighbors that are really close?
If there are too many points (for instance with a large area or with a small Tree Support Branch Distance), then it currently doesn't really converge to a smaller base.

Reply:
Ahhh the nodes were not removed from nodes_per_part, so they got inserted anyway in the loop below!

![cura-4523-before2](https://user-images.githubusercontent.com/1016604/34214262-39cbe99a-e5a2-11e7-858b-a0617247c5ee.png)
![cura-4523-after2](https://user-images.githubusercontent.com/1016604/34214265-3c254240-e5a2-11e7-9e02-b6966cb704f8.png)

Reply:
You had a fix but didn't push the commit when you went on vacation. In the meanwhile I also fixed it.